src/Uno.Toolkit.UI/Controls/Shadows/ShadowContainer.cs

```diff
@@ -0,0 +1,163 @@
+﻿using System;
+using System.Linq;
+
+#if IS_WINUI
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using SkiaSharp.Views.Windows;
+#else
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using SkiaSharp.Views.UWP;
+#endif
+
+#if ANDROID
+using Android.Views;
+#endif
+
+namespace Uno.Toolkit.UI;
+
+/// <summary>
+/// Provides the possibility to add many colored shadows to its content.
+/// </summary>
+/// <remarks>
+/// For now it renders badly on WASM due to a bug on the wasm skiasharp construction of the SKXamlCanvas.
+/// It should be fixed when this PR will be merged: https://github.com/mono/SkiaSharp/pull/2443
+/// </remarks>
+[TemplatePart(Name = nameof(PART_Canvas), Type = typeof(Canvas))]
+public partial class ShadowContainer : ContentControl
+{
+	private const string PART_Canvas = "PART_Canvas";
+
+	private Canvas? _canvas;
+
+#if false // ANDROID
+    private readonly SKSwapChainPanel _shadowHost;
+    private bool _notOpaqueSet = false;
+#else
+	private SKXamlCanvas? _shadowHost;
+#endif
+
+	private static readonly ShadowsCache Cache = new ShadowsCache();
+
+	private FrameworkElement? _currentContent;
+
+	private CornerRadius _cornerRadius;
+
+	public ShadowContainer()
+	{
+		_cornerRadius = new CornerRadius(0);
+	}
+
+	protected override void OnApplyTemplate()
+	{
+		_canvas = (Canvas)GetTemplateChild(nameof(PART_Canvas));
+
+
+#if false // ANDROID: We keep that as a reference cause it would
+        var skiaCanvas = new SKSwapChainPanel();
+        skiaCanvas.PaintSurface += OnSurfacePainted;
+#else
+		var skiaCanvas = new SKXamlCanvas();
+		skiaCanvas.PaintSurface += OnSurfacePainted;
+#endif
+
+#if __IOS__ || __MACCATALYST__
+        skiaCanvas.Opaque = false;
+#endif
+
+		_shadowHost = skiaCanvas;
+		_canvas.Children.Insert(0, _shadowHost!);
+
+		base.OnApplyTemplate();
+	}
+
+	/// <inheritdoc/>
+	protected override void OnContentChanged(object oldContent, object newContent)
+	{
+		if (oldContent is FrameworkElement oldElement)
+		{
+			_canvas?.Children.Remove(oldElement);
+			oldElement.SizeChanged -= OnContentSizeChanged;
+
+			// TODO unregister corner radius property changed
+		}
+
+		if (newContent is FrameworkElement newElement)
+		{
+			_currentContent = newElement;
+			_currentContent.SizeChanged += OnContentSizeChanged;
+
+			if (TryGetCornerRadius(newElement, out var cornerRadius))
+			{
+				// TODO register corner radius property changed
+			}
+
+			_cornerRadius = cornerRadius;
+		}
+
+		base.OnContentChanged(oldContent, newContent);
+	}
+
+	private static bool TryGetCornerRadius(FrameworkElement element, out CornerRadius cornerRadius)
+	{
+		CornerRadius? localCornerRadius = element switch
+		{
+			Control control => control.CornerRadius,
+			StackPanel stackPanel => stackPanel.CornerRadius,
+			RelativePanel relativePanel => relativePanel.CornerRadius,
+			Grid grid => grid.CornerRadius,
+			Border border => border.CornerRadius,
+			_ => null,
```

Comment:
could try to fallback on reflection:

```suggestion
			_ => VisualTreeHelperEx.TryGetDpValue<CornerRadius>(element, "CornerRadius", out var value) ? value : default(CornerRadius?),
```